modules/core/src/smithy4s/Endpoint.scala

```diff
@@ -63,75 +63,114 @@ object Endpoint {
     def errorable: Option[Errorable[E]]
     def streamedInput: StreamingSchema[SI]
     def streamedOutput: StreamingSchema[SO]
+    def builder: Builder[I, E, O, SI, SO] = Builder.fromEndpoint(this)
+  }
+
+  trait Middleware[A] { self =>
```

Comment:
`Endpoint.Middleware` can be generalised and hosted by the core module. It's actually very valuable to have access to it, as it allows to generalise the routing logic.